1169 sync engine with block cache trait

Oscar-Pepper opened the pull request:
- Builds on #1184, implementing an async wallet synchronization function
- Replaces `FsBlockDb` with `BlockCache` for generic implementation of block caches, solving cyclic dependencies.
- Implements `BlockCache` for `FsBlockDb` and `TestBlockCache` (formerly `BlockCache`)

str4d commented:
This needs to be rebased now that #1184 has been merged.

Oscar-Pepper commented:
> This needs to be rebased now that #1184 has been merged.

Ok thanks. I think its best I separate this PR into more managable chunks. I will try to fit this in soon so it doesnt get too far behind

str4d commented:
This PR has now been blocking parts of `zcashd` deprecation for almost 4 months. I'm going to do the rebase myself and open a new PR.

str4d commented on zcash_client_backend/src/data_api/chain.rs:
```diff
@@ -479,142 +472,167 @@ impl ScanSummary {
 /// This function will return after scanning at most `limit` new blocks, to enable the caller to
 /// update their UI with scanning progress. Repeatedly calling this function with `from_height ==
 /// None` will process sequential ranges of blocks.
-#[tracing::instrument(skip(params, block_source, data_db))]
+#[tracing::instrument(skip(params, block_cache, wallet_data))]
 #[allow(clippy::type_complexity)]
-pub fn scan_cached_blocks<ParamsT, DbT, BlockSourceT>(
+pub fn scan_cached_blocks<ParamsT, DbT, BcT>(
     params: &ParamsT,
-    block_source: &BlockSourceT,
-    data_db: &mut DbT,
-    from_height: BlockHeight,
-    limit: usize,
-) -> Result<ScanSummary, Error<DbT::Error, BlockSourceT::Error>>
+    block_cache: &BcT,
+    wallet_data: &mut DbT,
+    scan_range: &ScanRange,
+) -> Result<ScanSummary, Error<DbT::Error, BcT::Error>>
```
As a suggestion for future PRs: part of the reason why this PR is harder to rebase is because there are commits like this one that rename all of the arguments as well as changing the parameter name, which were not necessary to do. In particular, renaming `data_db` to `wallet_data` in the same commit as switching from `BlockSource` to `BlockCache`.

str4d commented on zcash_client_backend/src/sync.rs:
```diff
@@ -138,7 +123,7 @@ where
     // 7) Loop over the remaining suggested scan ranges, retrieving the requested data
     //    and calling `scan_cached_blocks` on each range.
-    let scan_ranges = db_data.suggest_scan_ranges()?;
+    let scan_ranges = wallet_data.suggest_scan_ranges()?;
```
Same thing here. The unnecessary rename from `db_data` to `wallet_data` has caused this entire file to be conflicted, but because it was done in the same commit as the `BlocksCache` change, I can't just blow the commit away and recreate the rename.

str4d commented on zcash_client_sqlite/src/testing.rs:
```diff
@@ -1352,104 +1349,94 @@ fn fake_compact_block_from_compact_tx(
-}
-
-pub(crate) struct BlockCache {
+pub(crate) struct TestBlockCache {
```
Another unnecessary rename. It's entirely expected to encounter structs with the same name as traits. Rather than renaming this struct in the same commit, this should instead have used a combination of `impl crate::data_api::chain::BlockCache for BlockCache` and `use crate::data_api::chain::BlockCache as _` to avoid the naming conflict and reduce the conflicting changes made in the commit.

zancas commented:
> This PR has now been blocking parts of `zcashd` deprecation for almost 4 months. I'm going to do the rebase myself and open a new PR.

Zingo Labs reconsidered its strategy and significantly reallocated resources with the April 30th release of Zashi.

str4d commented:
> Zingo Labs reconsidered its strategy and significantly reallocated resources with the April 30th release of Zashi.
>
> I have done my best to make my position clear and consistent.

Had you mentioned this here or elsewhere, I would have taken over the PR much earlier. As it was, I was trying to ensure you were included in the development process, something you had at the time expressed significant interest in, and was taking pains to ensure I was not doing work in parallel that would obsolete work you were presumably (given no updates here to the contrary) working on at your own pace.

I will not make this mistake in future.

zancas commented:
https://forum.zcashcommunity.com/t/zcashd-deprecation-needs-a-champion/48580/22

zancas commented:
I can check with `Zingoistas` maybe someone would be happy to pick this up.  Certainly we can add it to our work if there's compensation.   Maybe the ECC would like to contract with us.

str4d commented:
As I said above, I've already picked this up myself, due to its necessity for zcashd deprecation. I'm currently on PTO though and not working on it until at least end of next week, so if anyone else is likely to make progress in the next week they are more than welcome to do so.

Oscar-Pepper commented:
> As I said above, I've already picked this up myself, due to its necessity for zcashd deprecation. I'm currently on PTO though and not working on it until at least end of next week, so if anyone else is likely to make progress in the next week they are more than welcome to do so.

Hi str4d, I apologise for my part in the way this has turned out, it was not my intention to cause blockers on zcashd deprecation or cause you frustration. I was not aware of this until now as I missed your comment from three weeks ago.

I think zancas has already mentioned we are limited on resources and funds. The zip317 release turned out to be a lot more work than expected as we replaced a large % of our legacy code along with a massive increase in test coverage. I had to pull off sync to prioritise this work. I agree this should have been communicated better.

This PR was initally developped alongside #1192 . As i mentioned in #1192  and also on signal and discord, this PR was intended to replace #1184  so you wouldnt have to repeat the work i had already done to get #1184  building and ready for merge. At the time, this PR was complete and only draft because #1192  was still under review and had multiple change requests which lead to me rebasing and re-working this PR multiple times. Simplifying this rebasing was actually the reason why i squashed my commits which was one of your conerns.

I understand you were very busy with the zashi release but I feel the lack of communication at this time is also a factor here.

In this context, I hope you see that I wasnt so much renaming 'db_data', I was intending these names to be the first to merge and to better reflect the fact that this sync engine was also going to be used by wallets like zingo where the wallet data is not purely a database persistance layer like sqlite. In heinsight I should have left it alone.

I am away for the next week but I ***will*** prioritise this PR when i return if you have not got onto it before hand.

Oscar-Pepper commented:
superseded by #1535
